File: datediff.py
#!/usr/bin/env python3

import argparse

YEAR_MIN = 1901
YEAR_MAX = 2999

month_days_leap = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
month_days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

cumm_days = [0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334]
cumm_days_leap = [0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335]

'''
Function to check if the year is a leap year.
'''


def check_leap_year(year):
    if(year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
        return True
    return False


'''
Class that's used in the App to throw exceptions
and print the error.
'''


class AppException(argparse.ArgumentTypeError):
    def __init__(self, data):
        self.data = data

    def __str__(self):
        return repr(self.data)


'''
Class to initialize the date.

Instance Variables:
day, month, year, is_leap_year, since_origin

Methods:
form_valid_date: carry out the operations to fill the 
                instance variables
check_valid_month_year: Check if the month and Year are valid.
check_valid_day: Check if the day is valid.
num_days_since_origin: Calculate the day since 01/01/0001
'''


class NewDate():
    def __init__(self, date):
        self.form_valid_date(date)
        self.since_origin = self.num_days_since_origin()

    def form_valid_date(self, date):
        date_string = date.split("/")
        # Check if day, month, year fields are of the correct length
        if(len(date_string) != 3 or len(date_string[0]) != 2 or len(date_string[1]) != 2 or len(date_string[2]) != 4):
            raise AppException(
                "Please Make sure that the date is passed in the correct format. [DD/MM/YYYY]")
        date_list = []
        # Convert all three fields to integers
        for i in date_string:
            b = int(i)
            date_list.append(b)
        self.day, self.month, self.year = date_list
        self.is_leap_year = check_leap_year(self.year)
        # Check all fields are within the correct range
        valid_date = self.check_valid_month_year()
        if not valid_date:
            raise AppException(
                "Invalid Date. Please Make sure that the date passed is correct")

    def check_valid_day(self):
        # Take into account if the year is a leap year
        if self.is_leap_year:
            if self.day not in range(1, month_days_leap[self.month-1]):
                return False
        else:
            if self.day not in range(1, month_days[self.month-1]):
                return False
        return True

    def check_valid_month_year(self):
        # Check if the month and year are in the correct range
        if self.month not in range(1, 13):
            return False
        if self.year not in range(YEAR_MIN, YEAR_MAX+1):
            return False
        # Check if the day is valid
        valid_day = self.check_valid_day()
        if not valid_day:
            return False
        return True

    def num_days_since_origin(self):
        # Calculate the cummulative days in the year ranging up
        # to the current day in the year.
        # Dependent upon the year. If year is a leap year,
        # then an extra day is added
        if self.is_leap_year:
            present_day = self.day + cumm_days_leap[self.month - 1]
        else:
            present_day = self.day + cumm_days[self.month - 1]
        # Calculate the number of leap years before the present year
        numLeapYears = (self.year-1)//4 - \
            (self.year-1)//100 + (self.year-1)//400
        # return the number of days which is the sum of
        # number of days of the present year the number of leap years
        # before the present year as these will be the extra days
        # needed to be added to the (previous years)* 365
        return (self.year-1)*365 + numLeapYears + present_day - 1


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=f"Calculate the number of full days \
                                     elapsed between the experiment's start date and end date. \
                                     Enter valid dates between 01/01/{YEAR_MIN} and 31/12/{YEAR_MAX}.")
    parser.add_argument("date1", help="Experiment Start Date")
    parser.add_argument("hyphen", help="The hyphen", default="-")
    parser.add_argument("date2", help="Experiment End Date")
    args = parser.parse_args()
    # Initialize and Validate the input dates
    dd1 = NewDate(args.date1)
    dd2 = NewDate(args.date2)
    # Print out the result
    print(max(abs(dd2.since_origin - dd1.since_origin) - 1, 0))

File: datediff.go
package main

import (
	"os"
	"strconv"
	"strings"
)

const yearMIN int = 1901
const yearMAX int = 2999

var (
	monthDaysLeap = [...]int{31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}
	monthDays     = [...]int{31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}
	cummDays      = [...]int{0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334}
	cummDaysLeap  = [...]int{0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335}
)

// Creating the struct to represent the input dates
type dateType struct {
	day         int
	month       int
	year        int
	isLeapYear  bool
	sinceOrigin int
}

// Check if the day is valid
func (date *dateType) checkValidDay() bool {
	if date.isLeapYear {
		if date.day < 1 || date.day >= monthDaysLeap[date.month-1] {
			return false
		}
	} else {
		if date.day < 1 || date.day >= monthDays[date.month-1] {
			return false
		}
	}
	return true
}

// Check if the month and the year are valid
func (date *dateType) checkValidMonthYear() bool {
	if date.month < 1 || date.month > 12 {
		return false
	}
	if date.year < yearMIN || date.year > yearMAX {
		return false
	}
	validDay := date.checkValidDay()
	if !(validDay) {
		return false
	}
	return true
}

// Calculate the number of days from 01/01/0001
func (date *dateType) daysSinceOrigin() int {
	var presentDay int
	if date.isLeapYear {
		presentDay = date.day + cummDaysLeap[date.month-1]
	} else {
		presentDay = date.day + cummDays[date.month-1]
	}
	numLeapYears := (date.year-1)/4 - (date.year-1)/100 + (date.year-1)/400
	return (date.year-1)*365 + numLeapYears + presentDay - 1
}

// Check if the year is a leap year or not
func checkLeapYear(year int) bool {
	if (year%4 == 0 && year%100 != 0) || (year%400 == 0) {
		return true
	}
	return false
}

// Convert the string array into an integer array
func convertToInteger(date []string) [3]int {
	var date1List = [3]int{}
	k := 0
	for _, i := range date {
		j, err := strconv.Atoi(i)
		if err != nil {
			panic(err)
		}
		date1List[k] = j
		k++
	}
	return date1List
}

// Check if the number date passed as the input is valid or not
func validateStringLength(date []string) bool {
	if len(date) != 3 || len(date[0]) != 2 || len(date[1]) != 2 || len(date[2]) != 4 {
		return false
	}
	return true
}

func minMax(a int, b int) (int, int) {
	if a > b {
		return a, b
	}
	return b, a

}

func main() {
	// Make sure that the user inputs the correct number
	if len(os.Args) != 4 {
		panic("Insufficient arguments provided. \nPlease make sure the the arguments are in the format DD/MM/YYYY - DD/MM/YYYY.")
	}
	// We just want the dates
	date1 := os.Args[1]
	date2 := os.Args[3]

	// Split the dates based on the "/" separator
	date1ListString := strings.Split(date1, "/")
	date2ListString := strings.Split(date2, "/")

	// Validate the length of the date strings
	validString1 := validateStringLength(date1ListString)
	validString2 := validateStringLength(date2ListString)
	if !validString1 || !validString2 {
		panic("Please Enter the correct dates")
	}

	// Convert the string arrays to integer arrays
	var date1List = convertToInteger(date1ListString)
	var date2List = convertToInteger(date2ListString)

	// Initialize date1
	date1Type := dateType{
		day:        date1List[0],
		month:      date1List[1],
		year:       date1List[2],
		isLeapYear: checkLeapYear(date1List[2]),
	}
	// Initialize date2
	date2Type := dateType{
		day:        date2List[0],
		month:      date2List[1],
		year:       date2List[2],
		isLeapYear: checkLeapYear(date2List[2]),
	}

	// Check the validity of both the dates
	if !date1Type.checkValidMonthYear() || !date2Type.checkValidMonthYear() {
		panic("Invalid Dates. Please try again")
	}

	// Calculate the number of days since the origin
	date1Type.sinceOrigin = date1Type.daysSinceOrigin()
	date2Type.sinceOrigin = date2Type.daysSinceOrigin()

	// Print out the result
	maxVal, minVal := minMax(date2Type.sinceOrigin, date1Type.sinceOrigin)
	if maxVal-minVal == 0 {
		println(0)
	} else {
		println(maxVal - minVal - 1)
	}
}
